jgrapht-core/src/main/java/org/jgrapht/alg/tour/FarthestInsertionHeuristicTSP.java

```diff
@@ -0,0 +1,346 @@
+    {
+        E longestEdge = null;
+        double longestEdgeWeight = -1;
+        int n = graph.vertexSet().size();
+        allDist = new double[n][n];
+        for (var edge : graph.edgeSet()) {
+            V source = graph.getEdgeSource(edge);
+            V target = graph.getEdgeTarget(edge);
+            if (!source.equals(target)) {
+                int i = mapping.getVertexMap().get(source);
+                int j = mapping.getVertexMap().get(target);
+                if (allDist[i][j] == 0) {
+                    allDist[i][j] = allDist[j][i] = graph.getEdgeWeight(edge);
+                    if (longestEdgeWeight < allDist[i][j]) {
+                        longestEdgeWeight = allDist[i][j];
+                        longestEdge = edge;
+                    }
+                }
+            }
+        }
+        return longestEdge;
+    }
+
+
+    /**
+     * Find the index of the unvisited vertex which is farthest from the partially constructed tour.
+     *
+     * @param start The unvisited vertices start at index {@code start}
+     * @return the index of the unvisited vertex which is farthest from the partially constructed tour.
+     */
+    private int getFarthest(int start)
+    {
+        int n = distances.length;
+        int farthest = -1;
+        double maxDist = -1;
+        for (int i = start; i < n; i++) {
+            double dist = distances[i];
+            if (dist > maxDist) {
+                farthest = i;
+                maxDist = dist;
+            }
+        }
+        return farthest;
+    }
+
+    /**
+     * Initialize distances from the unvisited vertices to the initial subtour
+     *
+     * @param tour a partial tour with {@code initialSubtour} at the beginning
+     */
+    private void initDistances(int[] tour)
+    {
+        int n = mapping.getVertexMap().size();
+        int start = initialSubtour.getVertexList().size();
+        distances = new double[n];
+        Arrays.fill(distances, start, n, Double.POSITIVE_INFINITY);
+        for (int i = start; i < n; i++) {
+            for (int j = 0; j < start; j++) {
+                distances[i] = Math.min(distances[i], allDist[tour[i]][tour[j]]);
+            }
+        }
+    }
+
+    /**
+     * Update the distances from the unvisited vertices to the partially constructed tour
+     *
+     * @param v the last vertex added to the tour
+     * @param start the unvisited vertices start at index {@code start}
+     */
+    private void updateDistances(int v, int start)
+    {
+        for (int i = start; i < distances.length; i++) {
+            distances[i] = Math.min(allDist[v][i], distances[i]);
+        }
+    }
+
+    /**
+     * Swaps the two elements at the specified indices in the given double array.
+     *
+     * @param arr the array
+     * @param i the index of the first element
+     * @param j the index of the second element
+     */
+    public static void swap(double[] arr, int i, int j)
```

Comment:
Make this private

Reply:
What do you think about moving the swap methods to the `org.jgrapht.util.ArrayUtil` class?

Reply:
Yes that's better!

Reply:
Thanks so much for your feedback. Done.